controller/source/logging/logger.hpp

```diff
@@ -0,0 +1,61 @@
+#pragma once
+#include <iostream>
+#include <fstream>
+#include <mutex>
+#include <unordered_map>
+#include <filesystem>
+
+namespace controller {
+
+/**
+ * Singleton logger class to store the history of each pair in a different file.
+*/
+class logger {
+public:
+  static auto get_instance() -> logger* {
+    static logger history_logger;
+    return &history_logger;
+  }
+
+  void log(const query& query_args, const bool& result, const std::string& message = {}) {
+    
+    // lock the mutex corresponding to the key
+    std::lock_guard<std::mutex> lock(m_keys_to_mutexes[query_args.key()]);
+
+    // open the key's log file output stream in append only mode
+    std::ofstream log_file(log_file_path(query_args.key()), std::ios::app);
```

Comment:
looks like a performance killer

Reply:
As agreed, storing the `ofstreams` in a map now.